src/D2L.CodeStyle.Analyzers/Async/Generator/SyncGenerator.cs

```diff
@@ -0,0 +1,182 @@
+
+namespace D2L.CodeStyle.Analyzers.Async.Generator;
+
+[Generator]
+internal sealed partial class SyncGenerator : IIncrementalGenerator {
+	public void Initialize( IncrementalGeneratorInitializationContext context ) {
+		// Collect all the methods we want to generate sync versions for
+		// individually (for better incremental builds)
+		IncrementalValuesProvider<(MethodDeclarationSyntax, Compilation)> methodsToGenerate =
+			context.SyntaxProvider
+			   .CreateSyntaxProvider( predicate: IsInterestingLookingMethod, transform: ExtractSyntax )
+			   .Where( static m => m != null )
+			   .Combine( context.CompilationProvider )
+			   .WithComparerThatIgnoresCompilation()!;
+
+		// Do the generation per method
+		IncrementalValuesProvider<MethodGenerationResult> generatedMethods =
+			methodsToGenerate
+				.Select( GenerateSyncMethod )
+				// Filter out things we simply couldn't generate anything for.
+				.Where( static mr => mr.HasValue )
+				.Select( static ( mr, _ ) => mr!.Value );
+
+		// Collect and generate the output files
+		//
+		// Opportunity for better caching here -- changing one source file
+		// should only impact one output file. Currently we only generate the
+		// methods when their source file changed, but we re-collect all of
+		// that when any file with [GenerateSync] has changed.
+		IncrementalValuesProvider<FileGenerationResult> generatedFiles =
+			generatedMethods.Collect()
+				.SelectMany( GenerateFiles );
+
+		context.RegisterSourceOutput( generatedFiles, WriteFiles );
+	}
+
+	private static bool IsInterestingLookingMethod(
+		SyntaxNode node,
+		CancellationToken _
+	) {
+		if( node is not MethodDeclarationSyntax method ) {
+			return false;
+		}
+
+		// We only care about async methods
+		if( !method.Modifiers.Any( t => t.IsKind( SyntaxKind.AsyncKeyword ) ) ) {
+			return false;
+		}
+
+		// The ones we care about have to have a particular attribute -- so
+		// perk up when we see _any_ attributes (we don't have a SemanticModel
+		// in this method.)
+		return method.AttributeLists.Any();
+	}
+
+	private static MethodDeclarationSyntax? ExtractSyntax(
+		GeneratorSyntaxContext context,
+		CancellationToken cancellationToken
+	) {
+		var methodDeclaration = (MethodDeclarationSyntax)context.Node;
+
+		var attributes = methodDeclaration.AttributeLists
+			.SelectMany( al => al.Attributes );
+
+		foreach( var attribute in attributes ) {
+			// We have the semantic model now, so make sure this method has
+			// the right attribute.
+			if( IsGenerateSyncAttribute( context.SemanticModel, attribute, cancellationToken ) ) {
+				return methodDeclaration;
+			}
+		}
+
+		return null;
+	}
+
+	internal static bool IsGenerateSyncAttribute(
+		SemanticModel model,
+		AttributeSyntax attribute,
+		CancellationToken cancellationToken
+	) {
+		var attributeConstructorSymbol = model.GetSymbolInfo( attribute, cancellationToken ).Symbol as IMethodSymbol;
+
+		if( attributeConstructorSymbol == null ) {

```

Comment:
Visual studio's recommendation isn't horrible.  The `is not` definitely reads nice:

![image](https://user-images.githubusercontent.com/12101120/152930405-29903156-cf6b-4c8e-a5e0-4bdcc5679e60.png)

Reply:
Ya I'm cool with the english, I just haven't gotten it in my muscle memory yet